internal/rbd/nodeserver.go

```diff
@@ -1190,26 +1197,54 @@ func (ns *NodeServer) xfsSupportsReflink() bool {
 func (ns *NodeServer) NodeGetVolumeStats(
 	ctx context.Context,
 	req *csi.NodeGetVolumeStatsRequest) (*csi.NodeGetVolumeStatsResponse, error) {
-	var err error
-	targetPath := req.GetVolumePath()
-	if targetPath == "" {
-		err = fmt.Errorf("targetpath %v is empty", targetPath)
+	var (
+		imgInfo rbdImageMetadataStash
+		err     error
+	)
+	volumePath := req.GetStagingTargetPath()
+	if volumePath == "" {
+		if volumePath = req.GetVolumePath(); volumePath == "" {
+			return nil, status.Error(codes.InvalidArgument, "volume path must be provided")
+		}
+	} else {
+		volumePath = volumePath + "/" + req.GetVolumeId()
+		imgInfo, err = lookupRBDImageMetadataStash(volumePath)
+		if err != nil {
+			log.ErrorLog(ctx, "failed to find image metadata, at stagingPath: %s, err: %v", volumePath, err)
 
-		return nil, status.Error(codes.InvalidArgument, err.Error())
+			return nil, status.Error(codes.NotFound, err.Error())
+		}
 	}
 
-	stat, err := os.Stat(targetPath)
+	var volCondition *csi.VolumeCondition
+	stat, err := os.Stat(volumePath)
```

Comment:
`stat` is not a sufficient check for everything. In case a filesystem went read-only after a network outage, `stat` will still report success.

`stat` on a block-device will probably always succeed?

Reply:
### Problem with Writes:

**FS:**
Detecting if the accessMode is ReadOnly might not be possible with the info available at NodeGetVolumeStatsRequest{}, if we try to read that from mounts, then if the mounts turn into read-only we will get wrong info.

**Block:**
On the other side, we cannot write data into a block volume, because that could overwrite some important data.

### Problem with Reads:


**FS:**
We don't know which file to read unless we write something into the root dir of mount and read the same, as mentioned above writing data is not easy.

**Block:**
I think Reads on block volume can be done, without any hiccups, but again I just don't want to address 1/4 cases.

Reply:
@nixpanic please let me know if you think of something better?

Reply:
For RBD devices, hopefully some rbd command can check the status of the mapped device.

I do not know what would be suitable for good CephFS health checking.

I'd recommend to get in touch with the Ceph developers and ask for their recommendation. Ideally use their developers mailinglist, so that we can link to your request and their responses in follow-up PRs.

Reply:
I'm aware of anything similar to what you seem to be looking for here.  If the device/filesystem is mapped/mounted, it's expected to work (unless its client instance was blocklisted).  If the cluster is broken, unreachable, etc, we expect that to be a temporary condition and block I/O until it's resolved.  So from that POV, a mapped/mounted volume is always usable.

Reply:
Thanks @idryomov :+1:

Reply:
There are some quicks done in `posix_fs_health_check()` (from the [GlusterFS sources](https://github.com/gluster/glusterfs/blob/devel/xlators/storage/posix/src/posix-helpers.c)). XFS seems to behave different from ext4, so that probably needs some checking too.

Ideally read/write PVCs are tested by reading and writing, read-only PVCs by reading, only. It might make sense to write to a (hidden) file, store a timestamp and compare that after reading. Or just `stat()` the root of the filesystem. Anything you do needs a timeout, as the operation might be blocked in case network is unavailable or something. It is crucial for the `VOLUME_CONDITION` to be marked unhealthy, and not return a failure while requesting the condition.

Reply:
reading and writing cannot be done, unfortunately, please go through https://github.com/ceph/ceph-csi/pull/2833/files#r799183365 and see if that answers it.

Reply:
It should be possible to read the mount options,and it it is not read-only, you should be able to write to a file on the filesystem. Document the filename you are using well, so that users can find the name in the documentation.

Without writing to a file, and reading the contents (timestamp?) afterwards, the checks would not be very useful as only few issues would be detected.

Users consuming block-mode volumes are on their own regarding this, sadly.

Reply:
There is also a move to use a subdir for the actual PV, and not the root of the filesystem. That gives the CSI-driver an option to store metadata (for fscrypt) in the root of the volume, outside of the view of the consumers. Doing this for a health-check file makes sense too.

Reply:
@nixpanic I had opened a new issue to track improvements.
As discussed, let us start with a simple stat() and we will improve it with https://github.com/ceph/ceph-csi/issues/3200 soon with another PR.